File: May-LeetCoding-Challenge/12-Single-Element-In-A-Sorted-Array/Single-Element-In-A-Sorted-Array.cpp
class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
        int left = 0, right = nums.size() - 1;
        while (left < right){
            int mid = (right - left) / 2 + left;
      
            if (nums[mid] == nums[mid + 1]){
                if ((mid - left) % 2 == 0) left = mid + 2;
                else right = mid - 1;
            }
            else if (nums[mid] == nums[mid-1]){
                if ((right - mid) % 2 == 0) right = mid - 2;
                else left = mid + 1;
            }
            else  return nums[mid];
        }
        return nums[left];
    }
};

File: May-LeetCoding-Challenge/12-Single-Element-In-A-Sorted-Array/Single-Element-In-A-Sorted-Array.kt
class SingleElementinaSortedArrayKotlin540 {
    fun singleNonDuplicate(nums: IntArray): Int {
        if (nums.size == 1) {
            return nums[0]
        }
        if (nums[0] != nums[1]) {
            return nums[0]
        }
        if (nums[nums.size - 1] != nums[nums.size - 2]) {
            return nums[nums.size - 1]
        }
        var left = 1
        var right = nums.size - 2
        while (left + 1 < right) {
            val mid = left + (right - left).shr(1)
            if (isSingle(nums, mid)) {
                return nums[mid]
            }
            when {
                singleInRight(nums, mid) -> left = mid
                else -> right = mid
            }
        }
        if (isSingle(nums, left)) {
            return nums[left]
        }
        return nums[right]
    }

    private fun isSingle(nums: IntArray, index: Int): Boolean = nums[index] != nums[index - 1] &&
            nums[index] != nums[index + 1]

    private fun singleInRight(nums: IntArray, index: Int): Boolean =
        (nums[index] == nums[index - 1] && index % 2 != 0) ||
                (nums[index] != nums[index - 1] && index % 2 == 0)
}